File: Labs/Lab07/lab07.rb
def char_to_num(c)
    if c =~ /[A-Za-z]/
      c.downcase.ord - 'a'.ord
    else
      -1
    end
  end
  
  def num_to_char(n, is_upper)
    is_upper ? ('A'.ord + n).chr : ('a'.ord + n).chr
  end
  
  def encrypt_line(line, key)
    result = ""
    key_index = 0
  
    line.each_char do |char|
      if char =~ /[A-Za-z]/
        text_value = char_to_num(char)
        key_value = char_to_num(key[key_index % key.length])
  
        encrypted_char = num_to_char((text_value + key_value) % 26, char == char.upcase)
        result += encrypted_char
        key_index += 1
      else
        result += char 
      end
    end
  
    result
  end
  
  puts "Please enter a key (letters only): "
  key = gets.chomp
  
  unless key =~ /^[A-Za-z]+$/
    puts "We're sorry, your key is not valid. Only letters are allowed."
    exit
  end
  
  puts "Please enter the filename to encrypt: "
  filename = gets.chomp
  
 
  begin
    input_file = File.open(filename, 'r')
    output_file = File.open("ECopy.txt", 'w')
  
    input_file.each_line do |line|
      encrypted_line = encrypt_line(line, key)
      output_file.puts encrypted_line
    end
  
    puts "The encryption is complete! Saved to 'ECopy.txt'."
  rescue Errno::ENOENT
    puts "Sorry.. It seems your file cannot be opened."
  end
  

File: Labs/Lab07/lab07.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <cctype>

int charToNum(char c) {
    return std::isupper(c) ? c - 'A' : std::tolower(c) - 'a';
}
char numToChar(int n) {
    return 'A' + n;
}

std::string encryptLine(const std::string& line, const std::string& key) {
    std::string result;
    int keyIndex = 0;

    for (char c : line) {
        if (std::isalpha(c)) {
            int textValue = charToNum(c);
            int keyValue = charToNum(key[keyIndex % key.length()]);
            result += numToChar((textValue + keyValue) % 26); 
            keyIndex++;
        } else {
            result += c;
        }
    }

    return result;
}

int main() {
    std::string key, filename;
    std::cout << "Please enter a key (Use letters only): ";
    std::cin >> key;
    for (char c : key) {
        if (!std::isalpha(c)) {
            std::cout << "This key is invalid. Only letters are allowed.\n";
            return 1;
        }
    }

    std::cout << "Enter the filename to encrypt: ";
    std::cin >> filename;

    std::ifstream inputFile(filename);
    if (!inputFile) {
        std::cout << "File could not be opened.\n";
        return 1;
    }
    std::ofstream outputFile("ECopy.txt");
    std::string line;
    while (std::getline(inputFile, line)) {
        outputFile << encryptLine(line, key) << "\n";
    }
    std::cout << "Encryption complete! Saved to 'ECopy.txt'.\n";
    return 0;
}
